src/lib.rs

```diff
@@ -701,47 +701,97 @@ impl AddAssign<Time> for Time {
+            );
+            self.0 += rhs.0.unsigned_abs();
+        }
     }
 }
 
 impl Add<Duration> for Time {
     type Output = Time;
 
     fn add(self, rhs: Duration) -> Self::Output {
-        debug_assert!(
-            isize::try_from(self.0).is_ok() && self.0.checked_add(rhs.0 as usize).is_some(),
-            "overflow detected: {self:?} + {rhs:?}"
-        );
-        Time(((self.0 as isize) + rhs.0) as usize)
+        if rhs.is_negative() {
+            debug_assert!(
+                self.0.checked_sub(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            Time(self.0 - rhs.0.unsigned_abs())
+        } else {
+            debug_assert!(
+                self.0.checked_add(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            Time(self.0 + rhs.0.unsigned_abs())
+        }
     }
 }
 
 impl Sub<Duration> for Time {
     type Output = Time;
 
     fn sub(self, rhs: Duration) -> Self::Output {
-        debug_assert!(
-            isize::try_from(self.0).is_ok() && (self.0 as isize).checked_sub(rhs.0).is_some(),
-            "overflow detected: {self:?} - {rhs:?}"
-        );
-        Time(((self.0 as isize) - rhs.0) as usize)
+        if rhs.is_negative() {
+            debug_assert!(
+                self.0.checked_add(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            Time(self.0 + rhs.0.unsigned_abs())
+        } else {
+            debug_assert!(
+                self.0.checked_sub(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            Time(self.0 - rhs.0.unsigned_abs())
+        }
+    }
+}
+
+impl SubAssign<Duration> for Time {
+    fn sub_assign(&mut self, rhs: Duration) {
+        if rhs.is_negative() {
+            debug_assert!(
+                self.0.checked_add(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            self.0 += rhs.0.unsigned_abs();
+        } else {
+            debug_assert!(
+                self.0.checked_sub(rhs.0.unsigned_abs()).is_some(),
+                "overflow detected: {self:?} + {rhs:?}"
+            );
+            self.0 -= rhs.0.unsigned_abs()
+        }
     }
 }
 
 impl Sub<Time> for Time {
     type Output = Duration;
 
     fn sub(self, rhs: Time) -> Self::Output {
-        debug_assert!(
-            isize::try_from(self.0).is_ok() && isize::try_from(rhs.0).is_ok(),
-            "overflow detected: {self:?} - {rhs:?}"
-        );
-        Duration(self.0 as isize - rhs.0 as isize)
+        if self > rhs {
+            debug_assert!(
+                isize::try_from(self.0 - rhs.0).is_ok(),
+                "Overflow detected, Time instances are too far apart: {self:?} - {rhs:?}"
+            );
+            Duration((self.0 - rhs.0) as isize)
```

Comment:
I think all we can do here is panic, or return a `Result`.

Reply:
Or bound the `Time` range to usize/2 . We'd need this check in every function, though, not just in the constructor. This also incurs a runtime cost.